akka-cluster-sharding/src/main/scala/akka/cluster/sharding/external/ExternalShardAllocationStrategy.scala

```diff
@@ -0,0 +1,219 @@
+
+    private val Key = ddataKey(typeName)
+
+    override def preStart(): Unit = {
+      log.debug("Starting ddata state actor for [{}]", typeName)
+      replicator ! Subscribe(Key, self)
+    }
+
+    var currentLocations: Map[ShardId, String] = Map.empty
+
+    override def receive: Receive = {
+      case c @ Changed(key: LWWMapKey[ShardId, String] @unchecked) =>
+        val newLocations = c.get(key).entries
+        currentLocations ++= newLocations
+        log.debug("Received updated shard locations [{}] all locations are now [{}]", newLocations, currentLocations)
+      case GetShardLocation(shard) =>
+        log.debug("GetShardLocation [{}]", shard)
+        val shardLocation = currentLocations.get(shard).map(asStr => AddressFromURIString(asStr))
+        sender() ! GetShardLocationResponse(shardLocation)
+      case GetShardLocations =>
+        log.debug("GetShardLocations")
+        sender() ! GetShardLocationsResponse(currentLocations.transform((_, asStr) => AddressFromURIString(asStr)))
+    }
+
+  }
+}
+
+class ExternalShardAllocationStrategy(systemProvider: ClassicActorSystemProvider, typeName: String)(
+    // local only ask
+    implicit val timeout: Timeout = Timeout(5.seconds))
+    extends ShardCoordinator.StartableAllocationStrategy {
+
+  private val system = systemProvider.classicSystem
+
+  import ExternalShardAllocationStrategy._
+  import akka.pattern.ask
+  import system.dispatcher
+
+  private val log = Logging(system, classOf[ExternalShardAllocationStrategy])
+
+  private var shardState: ActorRef = _
+
+  private[akka] def createShardStateActor(): ActorRef = {
+    system
+      .asInstanceOf[ExtendedActorSystem]
+      .systemActorOf(DDataStateActor.props(typeName), s"external-allocation-state-$typeName")
+  }
+
+  private val cluster = Cluster(system)
+
+  override def start(): Unit = {
+    shardState = createShardStateActor()
+  }
+
+  override def allocateShard(
+      requester: ShardRegion,
+      shardId: ShardId,
+      currentShardAllocations: Map[ShardRegion, immutable.IndexedSeq[ShardId]]): Future[ShardRegion] = {
+
+    log.debug("allocateShard [{}] [{}] [{}]", shardId, requester, currentShardAllocations)
+
+    // current shard allocations include all current shard regions
+    (shardState ? GetShardLocation(shardId))
+      .mapTo[GetShardLocationResponse]
+      .map {
+        case GetShardLocationResponse(None) =>
+          log.debug("No specific location for shard [{}]. Allocating to requester [{}]", shardId, requester)
+          requester
+        case GetShardLocationResponse(Some(address)) =>
+          // if it is the local address, convert it so it is found in the shards
+          if (address == cluster.selfAddress) {
+            currentShardAllocations.keys.find(_.path.address.hasLocalScope) match {
+              case None =>
+                log.warning("unable to find local shard in currentShardAllocation. Using requester")
```

Comment:
not sure that is a warning, debug or info

Reply:
I wasn't sure how it could happen, hence the warning.

But maybe it will for the first allocation. Dropping to DEBUG.